src/java.base/share/native/libjli/java.c

```diff
@@ -537,8 +543,8 @@ JavaMain(void* _args)
     }
 
     /* If the user specified neither a class name nor a JAR file */
-    if (printXUsage || printUsage || what == 0 || mode == LM_UNKNOWN) {
-        PrintUsage(env, printXUsage);
+    if (printUsageKind != HELP_NONE) {
```

Comment:
Is the state check for `what` and `mode` no longer needed here?

Reply:
The idea was that `printUsageKind` should replace that - specifically, `HELP_CONCISE` should be equivalent to the original `what == 0 || mode == LM_UNKNOWN`. The code is a bit cryptic, so I hope I managed to achieve that.